File: backend/app/models/schedule.py
"""Lịch công tác — ScheduleItem + ScheduleReminder."""
from __future__ import annotations

from datetime import date, datetime, time
from typing import Optional

from sqlalchemy import (
    Boolean, Date, DateTime, ForeignKey, Integer,
    String, Text, Time, func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ScheduleItem(Base):
    __tablename__ = "schedule_items"

    id: Mapped[int]          = mapped_column(Integer, primary_key=True, index=True)
    leader_id: Mapped[int]   = mapped_column(ForeignKey("staff.id", ondelete="CASCADE"), nullable=False, index=True)

    # Nội dung lịch
    title: Mapped[str]           = mapped_column(String(500), nullable=False)
    location: Mapped[str | None] = mapped_column(String(300))
    note: Mapped[str | None]     = mapped_column(Text)

    # Thời gian
    work_date: Mapped[date]      = mapped_column(Date, nullable=False, index=True)
    # session: sang / chieu / ca_ngay / toi
    session: Mapped[str]         = mapped_column(String(20), nullable=False, default="sang")
    start_time: Mapped[time | None] = mapped_column(Time)

    # Nhắc Zalo
    zalo_remind: Mapped[bool]           = mapped_column(Boolean, default=False, nullable=False)
    remind_before_minutes: Mapped[int]  = mapped_column(Integer, default=30, nullable=False)

    # Quản lý
    created_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    created_at: Mapped[datetime]   = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), onupdate=func.now())
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), index=True)

    # Relationships
    leader   = relationship("Staff", foreign_keys=[leader_id], lazy="select")
    creator  = relationship("User", foreign_keys=[created_by], lazy="select")
    reminders: Mapped[list["ScheduleReminder"]] = relationship(
        "ScheduleReminder", back_populates="schedule", cascade="all, delete-orphan"
    )


class ScheduleReminder(Base):
    __tablename__ = "schedule_reminders"

    id: Mapped[int]          = mapped_column(Integer, primary_key=True, index=True)
    schedule_id: Mapped[int] = mapped_column(
        ForeignKey("schedule_items.id", ondelete="CASCADE"), nullable=False, index=True
    )
    leader_id: Mapped[int]   = mapped_column(Integer, nullable=False)

    zalo_user_id: Mapped[str | None] = mapped_column(String(100))
    recipient_phone: Mapped[str | None] = mapped_column(String(20))

    scheduled_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)
    sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    # pending / sent / failed / skipped
    status: Mapped[str] = mapped_column(String(20), default="pending", nullable=False, index=True)
    error_msg: Mapped[str | None] = mapped_column(Text)
    zalo_msg_id: Mapped[str | None] = mapped_column(String(100))
    retry_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    schedule: Mapped["ScheduleItem"] = relationship("ScheduleItem", back_populates="reminders")

File: frontend/src/types/schedule.ts
export type ScheduleSession = 'sang' | 'chieu' | 'ca_ngay' | 'toi'
export type ReminderStatus = 'pending' | 'sent' | 'failed' | 'skipped'

export const SESSION_LABELS: Record<ScheduleSession, string> = {
  sang:    'Sáng',
  chieu:   'Chiều',
  ca_ngay: 'Cả ngày',
  toi:     'Tối',
}

export const SESSION_COLORS: Record<ScheduleSession, string> = {
  sang:    'bg-amber-100 text-amber-800',
  chieu:   'bg-blue-100 text-blue-800',
  ca_ngay: 'bg-green-100 text-green-800',
  toi:     'bg-indigo-100 text-indigo-800',
}

export const REMIND_OPTIONS = [
  { value: 15,   label: '15 phút trước' },
  { value: 30,   label: '30 phút trước' },
  { value: 60,   label: '1 giờ trước' },
  { value: 120,  label: '2 giờ trước' },
  { value: 1440, label: '1 ngày trước' },
]

export interface LeaderMin {
  id: number
  full_name: string
  position: string | null
  employee_code: string | null
}

export interface ScheduleItemRead {
  id: number
  leader_id: number
  leader: LeaderMin | null
  title: string
  location: string | null
  note: string | null
  work_date: string          // YYYY-MM-DD
  session: ScheduleSession
  start_time: string | null  // "HH:MM:SS"
  zalo_remind: boolean
  remind_before_minutes: number
  created_by: number | null
  created_at: string
  updated_at: string | null
}

export interface ScheduleItemCreate {
  leader_id: number
  title: string
  location?: string
  note?: string
  work_date: string
  session: ScheduleSession
  start_time?: string        // "HH:MM"
  zalo_remind: boolean
  remind_before_minutes: number
}

export interface ScheduleItemUpdate {
  leader_id?: number
  title?: string
  location?: string
  note?: string
  work_date?: string
  session?: ScheduleSession
  start_time?: string
  zalo_remind?: boolean
  remind_before_minutes?: number
}

export interface ReminderLog {
  id: number
  schedule_id: number
  leader_id: number
  zalo_user_id: string | null
  scheduled_at: string
  sent_at: string | null
  status: ReminderStatus
  error_msg: string | null
  retry_count: number
  created_at: string
}

export interface PaginatedSchedule {
  items: ScheduleItemRead[]
  total: number
  page: number
  size: number
  pages: number
}

export interface WeekDay {
  [dateKey: string]: ScheduleItemRead[]
}

export interface WeekLeaderRow {
  leader: LeaderMin
  days: WeekDay
}

export interface WeekView {
  week_start: string
  week_end: string
  days: string[]
  leaders: WeekLeaderRow[]
}
